Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -0,0 +1,188 @@
+  ext; simp
+
+theorem asIdeal_matricesOver [DecidableEq n] (I : TwoSidedIdeal R) :
+    asIdeal (I.matricesOver n) = (asIdeal I).matricesOver n := by
+  ext; simp
+
+variable {n : Type*} [Fintype n] [DecidableEq n]
+
+/--
+Two-sided ideals in `R` correspond bijectively to those in `Mₙ(R)`.
+Given an ideal `I ≤ R`, we send it to `Mₙ(I)`.
+Given an ideal `J ≤ Mₙ(R)`, we send it to `{x₀₀ | x ∈ J}`.
+-/
+@[simps]
+def equivMatricesOver (o : n) : TwoSidedIdeal R ≃ TwoSidedIdeal (Matrix n n R) where
+  toFun I := I.matricesOver n
+  invFun J := TwoSidedIdeal.mk'
+    { x : R | ∃ N ∈ J, x = N o o }
+    ⟨0, J.zero_mem, rfl⟩
+    (by rintro _ _ ⟨x, hx, rfl⟩ ⟨y, hy, rfl⟩; exact ⟨x + y, J.add_mem hx hy, rfl⟩)
+    (by rintro _ ⟨x, hx, rfl⟩; exact ⟨-x, J.neg_mem hx, rfl⟩)
+    (by
+      rintro x _ ⟨y, hy, rfl⟩
+      exact ⟨diagonal (fun _ ↦ x) * y, J.mul_mem_left _ _ hy, by simp⟩)
+    (by
+      rintro _ y ⟨x, hx, rfl⟩
+      exact ⟨x * diagonal (fun _ ↦ y), J.mul_mem_right _ _ hx, by simp⟩)
+  right_inv J := SetLike.ext fun x ↦ by
+    simp only [mem_mk', Set.mem_image, SetLike.mem_coe, mem_matricesOver]
+    constructor
+    · intro h
+      choose y hy1 hy2 using h
+      rw [matrix_eq_sum_stdBasisMatrix x]
+      refine sum_mem fun i _ ↦ sum_mem fun j _ ↦ ?_
+      suffices
+          stdBasisMatrix i j (x i j) =
+          stdBasisMatrix i o 1 * y i j * stdBasisMatrix o j 1 by
+        rw [this]
+        exact J.mul_mem_right _ _ (J.mul_mem_left _ _ <| hy1 _ _)
+      ext a b
+      by_cases hab : a = i ∧ b = j
+      · rcases hab with ⟨ha, hb⟩
+        subst ha hb
+        simp only [stdBasisMatrix, and_self, ↓reduceIte, StdBasisMatrix.mul_right_apply_same,
+          StdBasisMatrix.mul_left_apply_same, one_mul, mul_one]
+        exact hy2 a b
+      · conv_lhs =>
+          dsimp [stdBasisMatrix]
+          rw [if_neg (by tauto)]
+        rw [not_and_or] at hab
+        rcases hab with ha | hb
+        · rw [mul_assoc, StdBasisMatrix.mul_left_apply_of_ne (h := ha)]
+        · rw [StdBasisMatrix.mul_right_apply_of_ne (hbj := hb)]
+    · intro hx i j
+      refine ⟨stdBasisMatrix o i 1 * x * stdBasisMatrix j o 1,
+        J.mul_mem_right _ _ (J.mul_mem_left _ _ hx), ?_⟩
+      rw [StdBasisMatrix.mul_right_apply_same, StdBasisMatrix.mul_left_apply_same,
+        mul_one, one_mul]
+  left_inv I := SetLike.ext fun x ↦ by
+    simp only [mem_mk', Set.mem_image, SetLike.mem_coe, mem_matricesOver]
+    constructor
+    · intro h
+      choose y hy1 hy2 using h
+      exact hy2 ▸ hy1 _ _
+    · intro h
+      exact ⟨of fun _ _ => x, by simp [h], rfl⟩
+
+@[simps!]
+def orderIsoMatricesOver (o : n) : TwoSidedIdeal R ≃o TwoSidedIdeal (Matrix n n R) := {
+  equivMatricesOver o with
+  map_rel_iff' := @fun I J => by
+    simp only [equivMatricesOver_apply]
+    constructor
+    · intro le x xI
+      specialize @le (of fun _ _ => x) (by simp [xI])
+      letI : Inhabited n := ⟨o⟩
+      simpa using le
+    · intro IJ M MI i j
+      exact IJ <| MI i j
+}
```

Comment:
```suggestion
def orderIsoMatricesOver (o : n) : TwoSidedIdeal R ≃o TwoSidedIdeal (Matrix n n R) where
  __ := equivMatricesOver o
  map_rel_iff' {I J} := by
    simp only [equivMatricesOver_apply]
    constructor
    · intro le x xI
      specialize @le (of fun _ _ => x) (by simp [xI])
      letI : Inhabited n := ⟨o⟩
      simpa using le
    · intro IJ M MI i j
      exact IJ <| MI i j
```
or similar

Reply:
I saw the `__ :=` elsewhere, but there is no hover-docstring so I don't know why this is preferrable. Could you point me to some documentation?

Reply:
This is just a style concern; `__` is semanticaly identical to `with`, but in general `where` notation is preferred for structures where possible.